[#46] Feat: 짤 업로드 api, 추천 태그 api 연결 및 Toast 구현

HY219 opened the pull request:
## 📝 작업 내용
- **태그 추천 및 검색 창 추가**

  - `RecommendTag 공통 컴포넌트` 이용

  - 추천 태그 목록과 일치하는 제목과 목록을 출력할 수 있도록, `RecommendTag 공통 컴포넌트`에 `title, recommendTags`를 props로 전달

- **태그와 이미지를 포함한 데이터를 서버에 전송 ( POST 요청 )**
  
  - 업로드 버튼 클릭 시, `useMutation`으로 POST 요청

  - formData로 한번에 데이터 전달

    > file: File
tagIdList: Array<number>
title: string

    > "Content-Type": "multipart/form-data"

  - `TODO: [2024.02.27] 실제 선택한 태그 목록 전송은, 선택한 태그의 Id를 전달하는 코드 구현 후 추가 작업할 예정입니다.`
  
- **전체 사용자들이 가장 많이 사용한 인기 태그 5개 출력 ( GET 요청 )**

  - `useQuery` 사용

- **사진, 태그 등록 여부, HTTP 요청 실패/성공 상태에 따라 Toast 메시지가 ~~5초간~~ 출력되도록 구현**
  - **`react-toastify` 이용**
  - **`toast.error` `toast.success` 이용**
  - **`toast.success`의 color를 primary와 동일하게 지정**

  - 업로드 완료 시, 업로드 완료 안내 및 업로드한 짤 페이지로 이동 버튼 토스트 구현

  - HTTP 요청 에러로 인한 사진 업로드 실패 시, 사진 업로드 실패 안내 토스트 구현
  
    - `mutation의 상태`에 따라 toast 출력
`isError` : HTTP 요청 실패
`isSuccess` : 짤 업로드 성공

  - 업로드 버튼 클릭 시 사진을 등록하지 않았을 경우, 사진 등록 안내 토스트 구현

  - 업로드 버튼 클릭 시 태그를 등록하지 않았을 경우, 태그 등록 안내 토스트 구현
    - 전역 상태로 선언되어있는 `selectedTags`로 태그 등록 여부 확인

- ~~**사진파일 등록 여부 상태가 변경되었을 때, Toast 출력 타이머 초기화**~~

  - ~~타이머가 만료되지 않았는데 `업로드 버튼`을 클릭할 경우, 새로 뜬 Toast가 기존 타이머의 남은 시간만큼만 출력되는 문제 해결~~

- **업로드 완료 시, 짤 업로드 페이지 초기화**

  - `file`과 전역의 `previewUrl`을 `null`로 두어 가져온 이미지 초기화

  - 전역의 `selectedTags`를 `[]`로 두어 선택한 태그 초기화

+previewUrl 상태로 관리하는 부분 제거 및 리팩토링

## 📷 스크린샷 (선택)
- 전체 사용자들이 가장 많이 사용한 인기 태그 5개 출력 ( GET 요청 )
  ![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/98df42a4-32cf-4448-aa9c-12a42e313db6)

- 업로드 완료 시, 업로드 완료 안내 및 업로드한 짤 페이지로 이동 버튼 토스트 구현
  ![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/e9c82a96-9bea-4fb3-af78-fd22966214ea)

- HTTP 요청 에러로 인한 사진 업로드 실패 시, 사진 업로드 실패 안내 토스트 구현
  ![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/6f553a91-21eb-4f09-92d1-82c3b243b033)


- 업로드 버튼 클릭 시 사진을 등록하지 않았을 경우, 사진 등록 안내 토스트 구현
  ![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/0d750f64-dd12-48d7-ab6c-7f0968f7f5d5)

- 업로드 버튼 클릭 시 태그를 등록하지 않았을 경우, 태그 등록 안내 토스트 구현
  ![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/69369df9-e10f-49f6-9e37-0373987c169f)


## 💬 리뷰 요구사항(선택)

- 변수, 함수, 폴더명이 올바른지, 더 괜찮은 이름이 있는지 확인 부탁드려요!
- 리팩토링이 필요한 부분을 알려주세요!
- ~~사용에 더 적합한 hook이 있다면 알려주세요!~~
- 리팩토링 예정
  - ~~가독성을 위해 `Toast 컴포넌트를 분리` 예정~~
  - ~~경고에 적합한 명칭을 위해 `delete` 와 같은 색을 가진 다른 명칭을 tailwind.config.js 파일에 추가 예정~~
  - 화면에 맞게 태그 검색 창 너비 조절 가능하도록 수정 예정 -> 머지 이후 새 이슈로 작업
  - ~~mutation의 상태에 따른 toast 출력 코드를 `usePostUploadZzal`폴더에 작성하도록 수정할 예정~~
  - 주석 삭제

## 📍 기타 (선택)

- 태그 검색과 자동완성 병합 구현이 완료된 후에 구현 예정인 작업들이 있습니다.
  - 병합된 태그 input 창 추가 + 주석 삭제
  - 실제 선택한 태그 아이디 목록을 전송
  - ~~태그 선택이 변경되었을 때, Toast 출력 타이머 초기화~~

**~~리팩토링이 필요한 부분들이 보여서, 리팩토링 후에 코드 살펴봐주셔도 좋을 것 같습니다!!~~
태그 검색과 selectedTags Id를 전달 받는 코드는 따로 구현이 필요해서, 추후에 추가할 예정입니다 :)**
  
close #46

HY219 committed:
Feat: 공통 Toast 컴포넌트 구현

toastColor, toastMessage, includeButton으로 기본 Toast 구성
buttonMessage, buttonRedirectPath로 버튼이 있을 경우  경로 이동

HY219 committed:
Feat: 사진파일 등록 여부에 따른 토스트 메시지 구현

사진파일 등록 여부 상태가 변경되었을 때, Toast 출력 타이머 초기화
사진파일을 등록했을 경우, Toast에 업로드한 짤 페이지로 이동하는 버튼 출력

HY219 committed:
Feat: 태그 추천 및 검색 창 추가

추천 태그 목록과 일치하는 제목과 목록을 출력할 수 있도록, RecommendTag 공통 컴포넌트에 title, recommendTags를 props로 전달

HY219 committed:
Feat: 태그 미등록 시 태그 등록 알림 Toast 출력

전역 상태로 선언되어있는 selectedTags로 태그 등록 여부 확인

HY219 committed:
Refactor: Toast 공통 컴포넌트를 UploadZzal 컴포넌트 파일로 통합 및 삭제

HY219 committed:
Feat: 짤 업로드 api 통합 구현

useMutation으로 POST 요청
업로드 버튼 클릭 시 등록한 이미지와 태그를 POST 요청
서버 이슈 해결 후 재확인 필요

HY219 committed:
Feat: 전체 유저 인기 태그 api 통합 구현

useQuery로 인기태그 GET 요청 및 화면에 출력
서버 이슈 해결 후 재확인 필요

HY219 committed:
Fix: 변경된 짤 upload POST 요청 형식에 맞게 데이터 타입수정

formData로 file데이터를 File형식으로 전달
dto안에 tagIdList(숫자리스트)랑 title(문자열) 전달

HY219 committed:
Feat: 짤 업로드 완료 시 초기화 구현

전역의 previewUrl, selectedTags를 두어, 성공완료시 초기화
가져온 이미지와 선택한 태그 초기화

HY219 committed:
Fix: 변경된 짤 업로드 POST 요청 타입에 맞게 변경

formData로 한번에 데이터 전달
file: File
tagIdList: Array<number>
title: string

HY219 committed:
Feat: HTTP 요청 실패 시, 사진 업로드 실패 에러 Toast 메시지 출력

짤 업로드 성공 시, file 데이터가 남아있는 이슈 수정
mutation의 상태에 따라 toast 출력
isError : HTTP 요청 실패
isSuccess : 짤 업로드 성공

HY219 committed:
Fix: API 요청 형식에 맞게 url 변경

C:/Program Files/Git/v1 추가

HY219 committed:
Modify: 직접 구현한 Toast 코드를 react-tostify를 활용한 코드로 변경

toast.success toast.error 사용
짤 업로드 성공 시 Toast에 업로드한 페이지로 이동하는 버튼 추가

HY219 committed:
Style: toast.success color를 primary color로 지정

#246fff 대신 primary로 작성할 경우, 인식이 안되는 이슈로 Hex Color로 작성

HY219 committed:
Rename: 업로드 버튼의 클릭 이벤트 핸들러 함수 이름 변경

handleShowToast -> handleClickUploadButton

HY219 committed:
Refactor: handleUploadZzal 함수에서 mutation 객체 직접 사용하여 성공/실패 콜백 처리

HY219 committed:
Merge branch 'main' of https://github.com/zzalmyu/zzalmyu-frontend into #46/feature/upload-zzal-api-intergration

HY219 committed:
Fix: 변경된 routes 구조 및 경로에 맞게 업로드한 짤 페이지 경로 재설정

C:/Program Files/Git/my-uploaded-zzal/ -> /my-uploaded-zzal

HY219 committed:
Refactor: 변경된 파일 경로와 일치하도록 TagSearchForm import 경로 수정

HY219 committed:
Fix: 전체 인기 태그를 가져오는 API의 response 타입 오류 해결

GetPopularTagsResponse의 타입을 Tag[]로 정의한 변경사항에 맞게 변경
GetPopularTagsResponse[] -> GetPopularTagsResponse

thalassophilia commented on src/index.css:
```diff
@@ -18,3 +18,7 @@
   clip-path: polygon(0 0, 0 0, 0 0);
 }
+
+:root {
+  --toastify-color-success: #246fff;
+}
```
tailwind의 config를 사용해주시는게 좋을 것 같아요

HY219 replied:
`--toastify-color-success`는 toatify의 css 변수인데, css 변수를 재정의하려면 css 파일에서 가능한 것으로 이해했습니다.
혹시 방법이 있을까요?

아니면 css 변수를 재정의하는 방식을 사용하지 않아야하는 걸까요?

thalassophilia commented on src/hooks/api/tag/useGetPopularTags.ts:
```diff
@@ -0,0 +1,17 @@
+  });
+
+  return {
+    popularTags: data || [],
```
data가 없으면 빈 배열이 오는게 아닌가요?

HY219 replied:
맞는 것 같습니다
그런데 `|| []` 를 삭제하게 되면 error가 발생했습니다.
useQuery를 사용할 때, 받아오는 데이터가 undefined일 수 있는 경우를 처리를 해줘야하는 것 같았어요.

다른 코드에서는 빈배열을 추가해주지 않아도 에러가 안나는 것 같아서 조금 더 찾아보도록하겠습니다

혹시 방법을 아신다면 공유 부탁드릴게요!

HY219 replied:
`Type 'GetPopularTagsResponse | undefined' is not assignable to type 'Tag[]'.
  Type 'undefined' is not assignable to type 'Tag[]'.`
위와 같은 에러가 발생합니다.

글을 하나 발견했는데, 이와 같은 문제일까 싶어, isSuccess로 해결 한 번 시도 해보려는데 맞는 방법일까요?
![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/064bd0b2-d009-4319-82c4-7df8fc03d86c)

https://chunho.tistory.com/m/82

thalassophilia commented on src/hooks/api/tag/useGetPopularTags.ts:
```diff
@@ -0,0 +1,17 @@
+  return {
+    popularTags: data || [],
+    isLoading,
+    error,
```
mutate에서 하신 것처럼 rest로 내보내주시고, isPending 사용을 추천한다고 합니다
https://github.com/TanStack/query/discussions/6297

HY219 committed:
Merge branch 'main' of https://github.com/zzalmyu/zzalmyu-frontend into #46/feature/upload-zzal-api-intergration

thalassophilia commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
-  const handleChangeUpload = (changedFile: File | null) => setFile(changedFile);
+  const [file, setFile] = useState<File | null>(null);
+  const [selectedTags, setSelectedTags] = useAtom($selectedTags);
+  const [, setPreviewUrl] = useAtom($previewUrl);
```
아톰이 아니라 useState 상태로 두고 props로 넘겨주면 안되나요?

HY219 replied:
성공했을 때, `.\src\components\UploadZzal\ImageUpload.tsx`로 props 넘겨서 처리해보려 했는데 쉽지 않네요. 이 방식이 맞을까요? 한 번 더 시도해보고 말씀드리겠습니다ㅜ!

thalassophilia commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
+  const [, setPreviewUrl] = useAtom($previewUrl);
+
+  const { popularTags } = useGetPopularTags();
+  const popularTagsName = popularTags.map((popularTag) => popularTag.tagName);
```
RecommendTag 컴포넌트가 이름이 아니라 popularTag 자체를 props로 받으면 더 좋을 것 같아요

HY219 replied:
popularTag 자체를 props로 전달 후, RecommendTag 컴포넌트에서 tagName을 추출해 사용했습니다:)

thalassophilia commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
+  const { popularTags } = useGetPopularTags();
+  const popularTagsName = popularTags.map((popularTag) => popularTag.tagName);
+
+  const { uploadZzal } = usePostUploadZzal();
```
훅은 훅끼리 상태는 상태끼리 선언문이 모야져있으면 보기 더 편할 거 같습니다!

HY219 replied:
위치 변경해줬습니다~!

thalassophilia commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
+  };
+
+  const handleClickUploadButton = () => {
+    if (!file) {
```
if-else 구문이 많아질수록 가독성이 떨어져서 ealry return문을 사용해주시면 더 좋을 것 같아요!

HY219 replied:
가독성면에서 훨씬 좋은 것 같네요:) 좋습니다 👍

thalassophilia commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
+  };
+
+  const handleUploadZzal = () => {
+    if (file) {
+      uploadZzal(
+        {
+          file: file,
+          // TODO: [2024.02.27] 선택한 태그의 Id를 전달하는 코드 구현 후, 실제 selectedTags Id 넘겨주기
+          tagIdList: [2, 3, 4],
+          title: file.name.substring(0, file.name.indexOf(".")),
+        },
+        {
+          onSuccess: () => {
+            toast.success(
+              <div>
+                <span>성공적으로 업로드가 되었습니다.</span>
+                <Link to="/my-uploaded-zzal">
+                  <button className="m-1 rounded bg-primary p-1 text-sm text-white">
+                    업로드한 짤 페이지로 이동
+                  </button>
+                </Link>
+              </div>,
+            ),
+              setFile(null),
+              setPreviewUrl(null),
+              setSelectedTags([]);
+          },
```
이 부분 뎁스가 너무 깊어서 파일이 없는 경우에 그냥 return 처리 해주시면 더 좋을 것 같아요

HY219 replied:
좋습니다! early return을 잘 활용해주면 좋네요:)

thalassophilia commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
+          file: file,
+          // TODO: [2024.02.27] 선택한 태그의 Id를 전달하는 코드 구현 후, 실제 selectedTags Id 넘겨주기
+          tagIdList: [2, 3, 4],
+          title: file.name.substring(0, file.name.indexOf(".")),
```
이 처리는 왜 필요한지 여쭤봐도 될까요?

HY219 replied:
이 부분은 파일의 이름을 그대로 가져와서 title에 전달해 준다고 생각해서 위와 같이 작성했었습니다.

현재 바뀐 UI에서는 title을 유저가 입력하도록 되어있어서, input창 추가하고 코드 수정하도록 하겠습니다!

thalassophilia commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
+            toast.success(
+              <div>
+                <span>성공적으로 업로드가 되었습니다.</span>
+                <Link to="/my-uploaded-zzal">
```
`my-uploaded-zzals`로 수정해주세요!

HY219 replied:
네! 경로에 맞게 수정했습니다 :)

thalassophilia commented on src/components/UploadZzal/ImageUpload.tsx:
```diff
@@ -1,13 +1,15 @@
+import { useAtom } from "jotai";
 import { cn } from "@/utils/tailwind";
 import ZzalCard from "@/components/common/ZzalCard";
+import { $previewUrl } from "@/store/zzal";
 interface Props {
```
한 줄 띄어주세요!

thalassophilia commented on src/components/common/RecommendTag.tsx:
```diff
@@ -2,16 +2,19 @@ import { cn } from "@/utils/tailwind";
+const RecommendTag = ({ title, recommendTags, className }: Props) => {
   // TODO: [2024.02.14] api 요청 예정
-  const recommendTags = ["분노", "스트레스", "박명수", "직장인", "잠좀자자"];
+  // const recommendTags = ["분노", "스트레스", "박명수", "직장인", "잠좀자자"];
```
주석 삭제해주세요!

thalassophilia commented on src/components/common/RecommendTag.tsx:
```diff
@@ -2,16 +2,19 @@ import { cn } from "@/utils/tailwind";
   return (
     <div className={cn("mb-4", className)}>
-      <div className="mb-4 text-sm font-bold">내가 가장 많이 사용한 태그</div>
+      {/* <div className="mb-4 text-sm font-bold">내가 가장 많이 사용한 태그</div> */}
```
이것도 삭제해주세요!

thalassophilia commented on src/components/common/RecommendTag.tsx:
```diff
@@ -2,16 +2,19 @@ import { cn } from "@/utils/tailwind";
+      {/* <div className="mb-4 text-sm font-bold">내가 가장 많이 사용한 태그</div> */}
+      <div className="mb-4 text-sm font-bold">{title}</div>
       <ul className="ml-2 flex w-650pxr flex-wrap gap-2 after:block after:w-650pxr after:border-b after:border-black after:content-['']">
         {recommendTags.map((recommendTag, index) => (
```
tagId를 사용하면 index를 사용하여 키 할당하지 않아도 될 것 같아요!

HY219 replied:
popularTag를 그대로 props로 전달받은 후 , tagId를 사용해서 index대신 키로 지정해주었습니다:)

yjc2021 commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
-          <div className="w-full border-2 border-text-secondary">searchbar</div>
+        <div className="flex h-300pxr w-full flex-1 flex-col justify-between">
+          <div className="w-full">
+            <div className="float-right">
```
`float` 속성을 `flexbox`를 활용하여 대체할 수는 없을까요?

레이아웃을 잡는 용도로 `float` 사용을 지양하는 걸로 알고 있어서 질문 드립니다
[관련 아티클](https://css-tricks.com/is-css-float-deprecated/)

HY219 replied:
오 감사합니다:)
이후에 UI 브랜치 파서 작업할 때 변경해줘도 될까요? float 속성을 flexbox로 변경하도록하겠습니다!

thalassophilia replied:
이거 오래 걸리시는게 아니면 해당 pr에서 해결해주시면 더 좋을 것 같아요!

choi-ik commented on src/apis/tag.ts:
```diff
@@ -1,5 +1,15 @@
+} from "@/types/tag.dto";
+import http from "./core";
+
+export const getPopularTags = () => {
+  return http.get<GetPopularTagsResponse>({
+    url: `/v1/tag/popular`,
+  });
+};
 
```
이 부분 {} 없이 바로 return 시켜줘도 될 것 같네요!

HY219 replied:
return 키워드 제거해주었습니다 :)

choi-ik commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
 const UploadZzal = () => {
-  const [, setFile] = useState<File | null>(null);
-  const handleChangeUpload = (changedFile: File | null) => setFile(changedFile);
+  const [file, setFile] = useState<File | null>(null);
```
기본 값이 null이라 제네릭 내의 null 빼줘도 될 것 같아요!!

HY219 replied:
초기값이 null일 경우에는 null을 타입으로 정의해주지 않아도 되는 군요! 배워갑니다!

그런데 null 타입을 제거할 경우, `setFile(null)`로 file을 초기화할 때 문제가 되는 것 같습니다.
이 경우를 위해서 null 타입은 그대로 두어야 할 것 같은데 어떻게 생각하시나요?

choi-ik commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,23 +1,92 @@
+          title: file.name.substring(0, file.name.indexOf(".")),
+        },
+        {
+          onSuccess: () => {
```
이 부분 mutation 파일에서 선언해줘도 좋지 않을까요??

HY219 replied:
https://github.com/zzalmyu/zzalmyu-frontend/pull/76#discussion_r1513654720

#76 pr에서 용재님께서 말씀해주신 것을 참고해서 말씀드리면, 이 경우도 api 성공 여부에따라 UI가 변경되는 코드인 것 같아서 mutate함수에서 onSuccess를 명시하는 것이 좋은 방식이라 생각하는데 어떻게 생각하시나요?

choi-ik commented on src/store/zzal.ts:
```diff
@@ -0,0 +1,3 @@
+import { atom } from "jotai";
+
+export const $previewUrl = atom<string | null>(null);
```
이 부분도 기본값이 null이라 제네릭내의 null 없어도 될 것 같아요!

HY219 replied:
previewUrl 도 file과 동일하게, 업로드 성공 시 null로 초기화를 해줘서 null 타입을 제거할 경우 error가 발생할 것 같아요.

혹시 다른 방법이 있을까요?

HY219 committed:
Cleanup: 불필요한 주석, 코드 삭제 및 import 순서 정렬

return 키워드 제거: return 키워드 없이 return 가능
줄 간격 띄움
불필요한 주석 삭제
훅은 훅끼리 상태는 상태끼리 선언문을 모아 가독성을 높이도록  import 순서 정렬(커스텀 api 훅은 상단에 배치)

HY219 committed:
Refactor: RecommendTag 컴포넌트의 recommendTags props에 popularTagsName이 아닌 popularTags 전달
recommendTags의 타입을 Tag[]로 변경
Array<string> -> Tag[]
RecommendTag 컴포넌트가 전달받은 추천 인기 태그 목록에서 tagName을 추출하여 출력

HY219 committed:
Refactor: if-else 구문을 early return 문으로 변경

if-else 구문이 많아질수록 가독성이 떨어지는 문제 해결

HY219 committed:
Merge branch 'main' of https://github.com/zzalmyu/zzalmyu-frontend into #46/feature/upload-zzal-api-intergration

HY219 committed:
Fix: 업로드한 짤 페이지로 이동하는 경로에 맞게 해당 경로 수정

C:/Program Files/Git/my-uploaded-zzals -> /my-uploaded-zzal

HY219 committed:
Refactor: 파일이 없을 경우 바로 return

handleUploadZzal()의 코드 뎁스가 깊어, 한 단계 뎁스 감소

HY219 committed:
Refactor: tagId를 사용하여 추천 태그의 키 할당

기존에 index를 사용해서 키를 할당한 부분 제거

HY219 committed:
Refactor: previewUrl을 atom이 아닌 useState 상태로 두고 props로 넘겨줘서 사용

HY219 committed:
Cleanup: previewUrl atom 생성 폴더 삭제 및 주석 삭제

HY219 committed:
Refactor: previewUrl 상태 제거

file 상태로만 화면에 보여주는 이미지 관리

file props 넘겨받기

HY219 committed:
Refactor: changeFile 함수와 중복되는 onChange 함수 제거

HY219 committed:
Merge branch 'main' of https://github.com/zzalmyu/zzalmyu-frontend into #46/feature/upload-zzal-api-intergration

HY219 committed:
Feat: 짤 제목 input 창 추가 및 등록

짤 업로드 시 사용자가 직접 짤 제목을 입력하여 등록하도록 추가

HY219 committed:
Merge branch 'main' into #46/feature/upload-zzal-api-intergration

HY219 committed:
Style: 기본 및 반응형 UI 수정

ImageUpload 고정, input을 화면 크기에 따라 조정

HY219 committed:
Fix: 짤 제목이 없을 경우, toast 띄운 후 바로 return

return 키워드 추가
짤 제목 없음과, 태그 없음 toast가 같이 뜨는 이슈 해결

HY219 committed:
Style: 짤 업로드 버튼 & 가이드 UI 수정

max-width를 주어 일정 너비까지만 줄어들도록 수정

thalassophilia commented on src/components/UploadZzal/ImageUpload.tsx:
```diff
@@ -60,8 +59,7 @@ const ImageUpload = ({ onChange }: Props) => {
     if (dataTransfer) {
```
dataTransfer 존재하는지 한 번 체크해주시는 이유가 있나요? 타입이 DataTransfer로 찍혀서 검사해줄 필요가 있나 해서요 : )

HY219 replied:
이 부분은 저도 확인해보고, 용재님이 짜두셨던 코드라 한 번 여쭤보겠습니다:)

HY219 replied:
![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/3bd41db8-c957-4d99-bab1-915ab204af94)
mdn 예시에서도 따로 체크해주고 있진 않은 것 같아요 👀 
![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/517ade63-04d9-4650-862d-6be61156b4d7)

thalassophilia replied:
파일이 없으면 타입이 어떻게 나오나요?

HY219 replied:
files 속성이 비어있을 수 있다면 dataTransfer 존재를 체크해주지 않더라도, `files`의 존재는 한번 체크해줘야할까요? @Kim-Jaemin420 @yjc2021

HY219 replied:
드레그해서 file을 넣어 주었을 때만 함수가 동작을 해서 파일이 없을 때 타입을 체크하는 것을 시도했는데 방법을 모르겠어서 혹시 어떻게 알 수 있나요? 😹

HY219 replied:
`드래그 해서 file을 넣어주었을때 함수가 동작한다면 dataTransfer가 항상 truthy한 값일텐데 if문으로 걸러주는거 같아` -> 의견에 동의해 falsy한 값이 들어갈 경우를 체크해주는 코드를 제거해주었습니다:)

thalassophilia commented on src/components/UploadZzal/ImageUpload.tsx:
```diff
@@ -2,27 +2,26 @@ import { Fragment, useRef, useState, DragEvent, ChangeEvent } from "react";
     const { files } = event.target;
     if (files) {
       const [changedFile] = files;
       if (changedFile) {
-        onChange(changedFile);
-        setPreviewUrl(URL.createObjectURL(changedFile));
+        changeFile(changedFile);
       }
```
if문 안의 if문으로 들어가면 뎁스가 깊어져서 저는 개인적으로 없는 경우에 return 처리해주는게 좋더라구요 : )
```js
    if (!files) return;

    const [changedFile] = files;
    
    if (changedFile) {
      changeFile(changedFile);
    }
```

HY219 replied:
좋습니다 통일성있게 위와 같이 수정해주도록 하겠습니다~!

thalassophilia commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,21 +1,109 @@
+              <span>성공적으로 업로드가 되었습니다.</span>
+              <Link to="/my-uploaded-zzals">
+                <button className="m-1 rounded bg-primary p-1 text-sm text-white">
+                  업로드한 짤 페이지로 이동
```
이거 문구가 너무 긴거 같은데 `확인하기` 정도는 어떤가요? @yjc2021 님도 의견주시면 감사하겠습니다!

HY219 replied:
저는 `확인하기` 또는 `이동` 도 괜찮은 것 같아요!

yjc2021 replied:
`확인하기` 좋은 것 같습니다 🦖

thalassophilia commented on src/routes/upload-zzal/index.tsx:
```diff
@@ -1,21 +1,109 @@
+  };
+
+  const handleUploadZzal = (file: File) => {
+    if (!file) {
+      return;
+    }
```
file 타입이 `File | null`이 아니라 File이라서 falsty값 걸러줄 필요가 없을 것 같은데요? 맞나요?

HY219 replied:
오 넵 
함수 매개변수 file의 타입이 `File` 인 경우에는 null인 경우를 체크해줄 필요없네요! 감사합니다:)

thalassophilia commented on src/components/common/RecommendTag.tsx:
```diff
@@ -1,21 +1,23 @@
-          <li key={`${index}-${recommendTag}`} className="mb-2">
-            <TagBadge content={recommendTag} isClickable />
+        {recommendTags.map((recommendTag) => (
+          <li key={`${recommendTag.tagId}`} className="mb-2">
+            <TagBadge content={recommendTag.tagName} isClickable />
```
`recommendTag.`같은 것도 중복의 일종이라 디스트럭처링 해주세요! : )

HY219 committed:
Refactor: 매개변수의 타입이 File인 경우 불필요한 falsy값 처리 코드 제거

함수의 매개변수의 타입이 null 이 아닐 경우, null인 경우를 체크해줄 필요없다.
-> null인 값이 인자로 들어갈 수가 없다.
file 타입이 File|null이 아니라 File이 falsy값을 걸러줄 필요 없음

HY219 committed:
Refactor: 중복 recommendTag 객체를 디스트럭처링 처리

{tagId, tagName}으로 디스트럭처링하여 중복된 코드 사용 제거

HY219 committed:
Refactor: handleFileDrop 함수에 불필요한 dataTransfer alsy 값 체크 코드 제거

항상 file이 드래그 됐을 경우에만 동작하는 함수이기 때문에 falsy한 값을 갖을 일이 없다

HY219 committed:
Rename: 짤 업로드 성공 시 페이지 이동 버튼 text 변경

업로드한 짤 페이지로 이동 -> 확인하기

thalassophilia reviewed (approved):
LGTM!

HY219 committed:
Refactor: 업로드할 미리보기 이미지 img태그를 사용하여 보여주도록 변경 및 style 변경

용도가 구분되는 ZzalCard 컴포넌트 대신 img 태그 상용
height 고정 제거
min-height와 object-contain width를 지정해주어 이미지가 가려지는 곳 없이 보여지도록 style 수정

HY219 committed:
Fix: 짤 업로드 성공 시, 입력한 짤 제목 초기화

input에 값이 그대로 남아있는 이슈 해결

HY219 committed:
Refactor: 규칙에 맞게 커스텀 api 훅의 요청 함수 형식 변경

getPopularTags -> () => getPopularTags()
postUploadZzal -> ({ file, tagIdList, title }: Props) => postUploadZzal({ file, tagIdList, title })

yjc2021 reviewed (approved):
LGTM 🦕

hyeonjinan096 reviewed (approved):
수고하셨습니다👍👍
